### ISSUE TITLE
[cmake] Inconsistent behavior with the official version of cmake

### ISSUE THREAD

**Issue by JackBoosY** [scikit-build/cmake-python-distributions#221]
Hi guys,
I'm vcpkg maintainer.
Recently we received an issue about using cmake 3.22.2 which installed by pip (https://github.com/microsoft/vcpkg/issues/22947):
Using the following cmake code has wrong results:
```cmake
message(STATUS "CMake ${CMAKE_VERSION}")
set(file /home/cuhk/vcpkg/packages/expat_x64-linux/debug/lib/pkgconfig/expat.pc)
message(STATUS "file: ${file}")
cmake_path(GET file PARENT_PATH pkg_lib_search_path)
message(STATUS "pkg_lib_search_path: ${pkg_lib_search_path}")
```
Expected:
```
pkg_lib_search_path: /home/cuhk/vcpkg/packages/expat_x64-linux/debug/lib/pkgconfig
```
Acturally:
```
pkg_lib_search_path: /home/cuhk/vcpkg/packages/expat_x64-linux/debug/lib/pkgconfig/expat.pc
```

Then we report this to cmake side: https://gitlab.kitware.com/cmake/cmake/-/issues/23187#note_1141682.

After discussions with the cmake maintainers, we believe that the cmake 3.22.2 version in pip is different from the official 3.22.2 version, and the modification of this bug may not be incorporated in pip side.

So can anyone take a look at this issue?

Thanks.
Jack

**Comment by mayeut** [comment]
I started investigating this.

It seems this starting to happen when cibuildwheel default image moved from manylinux2010 to manylinux2014.
Rebuilding 3.22.2 on manylinux2010, this specific issue goes away (I've not run CMake test suite).
So it seems to be related to the toolchain available on manylinux2014 that doesn't play well with CMake source code.

Here are a few tests I've done:
manylinux2014_x86_64:2021-12-12-e5100b5
- CMake 3.22.2, gcc10, CMake test suite failing
manylinux2014_x86_64:latest (there's been an update on the gcc10 toolchain)
- CMake 3.22.2, gcc10, CMake test suite failing
- CMake 3.23.0-rc1, gcc10, CMake test suite failing
- CMake 3.22.2, gcc11, CMake test suite passing (specific bug reported here is gone)
- CMake 3.23.0-rc1, gcc11, CMake test suite passing (specific bug reported here is gone)
 
First, we should really be running CMake test suite (I don't know how if it's really doable for alt-arch built with qemu given the time limit)
Given alt-arch are only available using manylinux2014, getting the x86_64, i686, build back on manylinux2010 doesn't seem right.

I will continue some testing with different toolchain and update my findings here for now.

**Comment by mayeut** [comment]
Fixed in 3.22.3
